### PR TITLE
Implement empty string coercion for StringCollectionDeserializer

### PR DESCRIPTION
**Pull request by yawkat**
This fixes https://github.com/FasterXML/jackson-dataformat-xml/issues/513 .

@cowtowncoder how can I test this PR? I can't do it in jackson-databind, because the XML JsonParser impl is not available. Making a custom impl that behaves the same way as XML just for this test would be pretty complicated.

The test case is:

```java
public class StringCollectionDeserializerTest extends TestCase {
    @Test
    public void testEmptyXml() throws IOException {
        ObjectMapper mapper = new XmlMapper()
                .enable(DeserializationFeature.ACCEPT_SINGLE_VALUE_AS_ARRAY)
                ;
        List<?> strings = mapper.readValue("<items></items>", new TypeReference<List<String>>() {});
        Assert.assertEquals(0, strings.size());
    }
}
```

### CODE REVIEW

File: src/main/java/com/fasterxml/jackson/databind/deser/std/StringCollectionDeserializer.java [hunk 1]
```diff
@@ -304,6 +306,21 @@ public Object deserializeWithType(JsonParser p, DeserializationContext ctxt,
                 throw JsonMappingException.wrapWithPath(e, result, result.size());
             }
         }
+        if (value != null) {
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/databind/deser/std/StringCollectionDeserializer.java:309]: Hmmmh. What kind of value are we expecting here? `JsonToken.VALUE_STRING` was handled earlier wasn't it?

**Comment by cowtowncoder** [comment]
Right, unfortunately it'd have to be in `jackson-dataformat-xml` for reasons mentioned; cannot be here.
And then requires use of a later snapshot.

There is also `jackson-integration-tests`, but that's maybe bit more to guard against regressions.

File: src/main/java/com/fasterxml/jackson/databind/deser/std/StringCollectionDeserializer.java (hunk 1 above)
Comment (yawkat) [reply src/main/java/com/fasterxml/jackson/databind/deser/std/StringCollectionDeserializer.java:309]: - The method is invoked when the token is *not* `START_ARRAY`
- This code path is visited when `ACCEPT_SINGLE_VALUE_AS_ARRAY` is *enabled*

So the token here could be anything non-array that is supported by the `valueDes`. The `value` could then be null.

Other than the `null` check this code is copied from `CollectionDeserializer`. Though in that case it does these checks *before* going through the `valueDes` (it just does them on `parser.getText()` if the token is `VALUE_STRING`). Is that a better approach?

Comment (cowtowncoder) [reply src/main/java/com/fasterxml/jackson/databind/deser/std/StringCollectionDeserializer.java:309]: This to me just seems wrong... values returned by a deserializer (or `getNullValue()`) really should not be post-processed by logic like this: coercion, in particular, is to be applied to incoming tokens and not on something deserializer has already decoded.

So this just feels conceptually against design. I understand it may result in desired outcome, it just does not seem to get there in a way API should work.

Not sure how to proceed here to be honest.

Comment (yawkat) [reply src/main/java/com/fasterxml/jackson/databind/deser/std/StringCollectionDeserializer.java:309]: I'll take a look how CollectionDeserializer does it, I don't think it processes the getNullValue. Maybe I can replicate that.

**Comment by altro3** [comment]
@cowtowncoder Hi! Could you review this PR? I think, this is very important fix

**Comment by cowtowncoder** [comment]
Ok, first things first: I think a test is definitely required here, to check handling of empty String for `Collection<String>` when enabled/disabled. It also seems to me that code might not work at all, since the `VALUE_STRING` is handled around line `282` (unless "unwrap" is enabled):

```
        if (!canWrap) {
            if (p.hasToken(JsonToken.VALUE_STRING)) {
                return _deserializeFromString(p, ctxt);
            }
            return (Collection<String>) ctxt.handleUnexpectedToken(_containerType, p);
        }
```

Another question is whether this same handling should be done for `StringArrayDeserializer`.

But looking at code again, it seems that `_deserializeFromString()` might already allow this conversion?
It does call `_deserializeFromEmptyString()` in its turn, at least.

So: let's start by adding unit test(s) to see what happens.

Fwtw I think the idea is fine for inclusion in 2.14.

**Comment by altro3** [comment]
@yawkat Hi! Yeah, the same proble. Base issue you can see here: https://github.com/micronaut-projects/micronaut-jackson-xml/issues/175

**Comment by yawkat** [comment]
@cowtowncoder i found no straight-forward way to test this in jackson-databind because it relies on the details of the xml json parser. it would require making a parser that has the same behavior in jackson-databind, but it's pretty complicated.

**Comment by cowtowncoder** [comment]
@yawkat  Ok, are you saying that the behavior for JSON is already as expected for this case? That adding a test with coercion from empty/blank String into empty `List<String>` works when `ACCEPT_SINGLE_VALUE_AS_ARRAY` is enabled?

**Comment by yawkat** [comment]
@cowtowncoder long comment.

When `ACCEPT_SINGLE_VALUE_AS_ARRAY` is on and you want to parse a `List<String>`, you expect either a `START_ARRAY` or `VALUE_STRING` token. For JSON this is unambiguous, it would be either `[]` or `""` input. 

For XML however, both an empty array and an empty string are written as `<items/>`. The XML mapper resolves this ambiguity by treating the input as an empty string, and then adding a coercion from empty string to empty list. But `StringCollectionDeserializer` never does this coercion, so it incorrectly yields a list with a single empty string, when it should be an empty list.

This patch resolves the issue by adding this coercion check. The logic comes from CollectionDeserializer. If you try the same test case with a non-String input, e.g. a `List<StringBuilder>`, the coercion applies correctly. https://github.com/FasterXML/jackson-databind/blob/e0fd026b38e9f58e78d191ba3b5f03c5a743b1b4/src/main/java/com/fasterxml/jackson/databind/deser/std/CollectionDeserializer.java#L292-L327

---

But looking at this PR again, I don't think I understood the mechanism correctly. Because this works by coercion, it is possible after all to replicate this without the xml mapper, simply by configuring the coercions the same way as with xml. I also identified a problem with the patch: If the coercion is *not* present (default) then it will now produce an error from the coercion failure. But this is also an issue with the normal `CollectionDeserializer`, which should behave the same way as `StringCollectionDeserializer` (looks like this was broken in `CollectionDeserializer` by 49c744d36501a16062b8293b65a252274f9de38b). I think it should be fixed in both: If the coercion fails, the code should instead continue to attempt to wrap the item in a list.

---

I've pushed some test cases to this PR. They all deserialize the input `""` to a list:

- Either with or without the coercion `.setCoercion(CoercionInputShape.EmptyString, CoercionAction.AsEmpty)` configured (with coercion emulates the XML mapper, without coercion is the JSON mapper)
- Either the element type `String`, so that the test case hits the `StringCollectionDeserializer`, or to a `StringWrapper` class, so that it hits the normal `CollectionDeserializer`.

The end goal of this PR is that *without* the coercion, the test cases deserialize to a single-item list with an empty string, and *with* the coercion they deserialize to an empty list. 

Before this patch, `StringCollectionDeserializer` behaved correctly *without* coercion, and `CollectionDeserializer` behaved correctly *with* coercion, but they failed with the other case. With the initial patch I made, both behaved correctly *with* coercion, but `StringCollectionDeserializer` now fails *without* coercion present. I've now changed the PR to pass all test cases, by ignoring a coercion failure (it can still fail later in non-string handling for `CollectionDeserializer`).

(note that the test class also contains some other tests, eg for blank strings, and for empty lists)

**Comment by altro3** [comment]
@yawkat Hi! Did you see it?
```
[INFO] Running com.fasterxml.jackson.databind.convert.CoerceContainersTest
Error:  Tests run: 12, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.007 s <<< FAILURE! - in com.fasterxml.jackson.databind.convert.CoerceContainersTest
Error:  testScalarCollections(com.fasterxml.jackson.databind.convert.CoerceContainersTest)  Time elapsed: 0.003 s  <<< FAILURE!
junit.framework.AssertionFailedError: 
Expected an exception with one of substrings ([Cannot coerce empty String]): got one (of type com.fasterxml.jackson.databind.exc.MismatchedInputException) with message "Cannot deserialize value of type `java.util.ArrayList<java.lang.Double>` from String value (token `JsonToken.VALUE_STRING`)
 at [Source: (String)""""; line: 1, column: 1]"
	at com.fasterxml.jackson.databind.convert.CoerceContainersTest._verifyNoCoercion(CoerceContainersTest.java:166)
	at com.fasterxml.jackson.databind.convert.CoerceContainersTest.testScalarCollections(CoerceContainersTest.java:32)
```

**Comment by yawkat** [comment]
missed that failure, thanks! it's just a change in the error message though, imo it's fine.

**Comment by cowtowncoder** [comment]
Ok. I hope I am not too confused, but just to be sure: if coercion is blocked (there is explicit "fail" setting), there needs to exception for attempt.

Things do get tricky when "unwrapping" is allowed, esp. due to XML inability to know if something is an array marker.
But probably even without that, with  JSON.

**Comment by yawkat** [comment]
@cowtowncoder i think that even when coercion is blocked explicitly by the user, it should not always fail. 

Say you have the input `""` and deserialize it as `List<String>`, with `ACCEPT_SINGLE_VALUE_AS_ARRAY`. The deserializer now has two options, it can try to coerce `""` to a list, or it can simply parse it as a single-item list with an empty string, with no coercion happening.

imo, even if coercion is explicitly set to fail, it should still try the second path. It is still a valid deserialization after all, it just depends on which approach is attempted first. If one approach would fail, and the other one succeeds, it should always pick the successful one.

This is also the behavior of `CollectionDeserializer` in 2e06068ff88c23e00add364c37025437c7a57c23, so there is some precedent for this. You can see that `CollectionDeserializer` checked whether the coercion failed, and if it did, then it still tried the non-array code path. This behavior was introduced in 2e06068ff88c23e00add364c37025437c7a57c23, but immediately removed again in 49c744d36501a16062b8293b65a252274f9de38b for some reason (though the comment describing this remained).

**Comment by cowtowncoder** [comment]
@yawkat Right, what I mean was specific coercion from "" into empty List -- but not automatic-wrapping via ACCEPT_SINGLE_VALUE_AS_ARRAY which should succeed. Latter would not be considered coercion in this sense (although I agree that one could use different definition of what coercion means for sure).

As to change linked to, it looks to me like just refactoring, moving handling of coercion to different method, but not removing it? (unless that occurred by accident)

**Comment by yawkat** [comment]
@cowtowncoder if the coercion is disabled, and ACCEPT_SINGLE_VALUE_AS_ARRAY is off, there will still be an error. it'll just be different. That's the test that's failing right now that @altro3 mentioned: It doesn't give a coercion error anymore, it gives a new error that the singular value can't be deserialized to a list.

As for the change in 49c744d36501a16062b8293b65a252274f9de38b , you can see that before that commit, there was a check `act != CoercionAction.Fail`. If the coercion failed, it would fall through into `handleNonArray`, where further handling takes place, like the comment describes:

> // handleNonArray may successfully deserialize the result (if
> // ACCEPT_SINGLE_VALUE_AS_ARRAY is enabled, for example) otherwise it
> // is capable of failing just as well as _deserializeFromEmptyString.

However 49c744d36501a16062b8293b65a252274f9de38b removed that check, the `act` goes into `_checkCoercionFail`, and throws an exception. It never hits `handleNonArray`.

My PR is basically restoring the behavior prior to `49c744d36501a16062b8293b65a252274f9de38b`, and also extending it to `StringCollectionDeserializer` (which doesn't do coercion at all atm).

**Comment by cowtowncoder** [comment]
@yawkat Ok that sounds good wrt handling, I think I get it now. It'd probably be good to explicit mention that `CoercionAction.Fail` is sort of ignored. Although code of course does that too.
Also good that handling is unified across general Collection and specific String-collection deserializers.

As to error message... yeah, that's tricky. Not sure if there's any better way to indicate the issue.

I guess I could just merge this and change test for now.

**Comment by cowtowncoder** [comment]
Ok sounds good; I am fine with this.

There are couple of minor changes I'll request and then I can merge this.

File: src/test/java/com/fasterxml/jackson/databind/deser/std/EmptyStringAsSingleValueTest.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.Collections;
+import java.util.List;
+
+public class EmptyStringAsSingleValueTest {
+    private static ObjectMapper normalMapper() {
+        ObjectMapper mapper = new ObjectMapper();
+        mapper.enable(DeserializationFeature.ACCEPT_SINGLE_VALUE_AS_ARRAY);
```
Comment (cowtowncoder) [inline src/test/java/com/fasterxml/jackson/databind/deser/std/EmptyStringAsSingleValueTest.java:19]: Could you change this to use builder pattern (`JsonMapper.builder().enable(...).build()`) instead? That way it'll merge to 3.0 cleanly.

**Comment by yawkat** [comment]
ok thanks, i will do them tomorrow.

File: src/test/java/com/fasterxml/jackson/databind/deser/std/EmptyStringAsSingleValueTest.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        return mapper;
+    }
+
+    private static ObjectMapper coercionMapper() {
+        ObjectMapper mapper = normalMapper();
+        // same as XmlMapper
+        mapper.coercionConfigDefaults()
```
Comment (cowtowncoder) [inline src/test/java/com/fasterxml/jackson/databind/deser/std/EmptyStringAsSingleValueTest.java:26]: Similar thing here, wrt coercion config defaults from `CoerceToBooleanTest` something like:

```
    private final ObjectMapper MAPPER_INT_TO_EMPTY = jsonMapperBuilder()
            .withCoercionConfig(LogicalType.Boolean, cfg ->
                cfg.setCoercion(CoercionInputShape.Integer, CoercionAction.AsEmpty))
            .build();

```

(modified to settings here of course)

**Review by cowtowncoder — CHANGES_REQUESTED** [review]
Looks good, would just like `ObjectMapper` construction to be made using builder so merging is easier (3.0 has immutable ObjectMappers)

**Comment by cowtowncoder** [comment]
> ok thanks, i will do them tomorrow.

Perfect. Thank you for your patience, it has taken quite a bit for me to grok things.
But I hope this helps with XML module, aside from fixing inconsistencies with JSON (etc) handling anyway.

**Comment by cowtowncoder** [comment]
Actually, I can make changes myself after merging.

**Review by cowtowncoder — APPROVED** [review]

**Comment by yawkat** [comment]
oh i totally forgot about implementing this for arrays too like you mentioned above. i will make a second pr for that.
